src/gen-ts.ts

```diff
@@ -8,295 +8,184 @@
+  for (const element of doc.getFeatures({kind: 'element'})) {
+    handleElementOrBehavior(element, root);
   }
-  const {name: className, namespace: namespaceName} =
-      getNamespaceAndName(element.className);
-  if (namespaceName !== 'Polymer') {
-    // TODO: handle non-Polymer namespaces
-    return;
+  for (const behavior of doc.getFeatures({kind: 'behavior'})) {
+    handleElementOrBehavior(behavior, root);
   }
+}
 
-  const node = element.astNode;
-  let d = '';
-
-  if (node == null) {
-    process.stderr.write(`no AST node for ${element.className}\n`);
-    d += `${idt(indent)}class ${className} {\n`;
-  } else if (
-      node.type === 'ClassDeclaration' || node.type === 'ClassExpression') {
-    d += `${idt(indent)}class ${node.id!.name} extends ${
-        generate(node.superClass)} {\n`;
-  } else if (node.type === 'VariableDeclaration') {
-    // We can't output the variable initializer, so we union the mixins
-    let supers: string[] = [];
-    if (element.superClass) {
-      const superName =
-          getNamespaceAndName(element.superClass.identifier).name!;
-      supers.push(superName)
-    }
-    for (const mixin of element.mixins) {
-      const mixinName = getNamespaceAndName(mixin.identifier).name!;
-      supers.push(mixinName)
-    }
-    const type = supers.map((s) => `Constructor<${s}>`).join('&');
-
-    d += `${idt(indent)}const ${className}: ${type};\n`;
-    declarations.push(d);
-    return;
+/**
+ * Add the given Element or Behavior to the given TypeScript declarations
+ * document.
+ */
+function handleElementOrBehavior(
+    feature: analyzer.Element|analyzer.PolymerBehavior, root: ts.Document) {
+  let className, parent;
+  if (feature.className) {
+    const parts = feature.className.split('.');
+    className = parts[parts.length - 1];
+    parent = findOrCreateNamespace(root, parts.slice(0, -1));
+  } else if (feature.tagName) {
+    className = kebabToCamel(feature.tagName);
+    parent = root;
   } else {
-    process.stderr.write(`no AST node for ${element.className}\n`);
-    d += `${idt(indent)}class ${className} {\n`;
+    console.error('Could not find a name.');
+    return;
   }
 
-  for (const [, property] of element.properties) {
-    if (property.privacy === 'private' || property.inheritedFrom != null) {
-      continue;
+  const extendz = new Array<string>();
+  if (isPolymerElement(feature)) {
+    extendz.push('Polymer.Element');
+    for (const behavior of feature.behaviorAssignments) {
+      extendz.push(behavior.name);
     }
-    d += `${idt(indent)}  ${getVisibility(property)}${property.name}${
-        getTypeAnnotation(property, true)};\n`;
+  } else if (isElement(feature)) {
+    extendz.push('HTMLElement');
   }
 
-  for (const [, method] of element.methods) {
-    const methodText = genMethod(method, indent + 2);
-    if (methodText) {
-      d += methodText;
+  const properties = new Array<ts.Property>();
+  for (const [, property] of feature.properties) {
```

Comment:
Prefer `feature.properties.values()` for just getting the values, that way you avoid allocating the array for each results.

Reply:
Done.